server/sonar-server/src/main/java/org/sonar/server/permission/index/AuthorizationDao.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.permission.index;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.dbutils.DbUtils;
+import org.apache.commons.lang.StringUtils;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+
+import static org.sonar.db.DatabaseUtils.executeLargeInputs;
+import static org.sonar.db.DatabaseUtils.repeatCondition;
+
+/**
+ * No streaming because of union of joins -> no need to use ResultSetIterator
+ */
+public class AuthorizationDao {
+
+  public static final class Dto {
+    private final String projectUuid;
+    private final long updatedAt;
+    private final List<String> users = Lists.newArrayList();
+    private final List<String> groups = Lists.newArrayList();
+
+    public Dto(String projectUuid, long updatedAt) {
+      this.projectUuid = projectUuid;
+      this.updatedAt = updatedAt;
+    }
+
+    public String getProjectUuid() {
+      return projectUuid;
+    }
+
+    public long getUpdatedAt() {
+      return updatedAt;
+    }
+
+    public List<String> getUsers() {
+      return users;
+    }
+
+    public Dto addUser(String s) {
+      users.add(s);
+      return this;
+    }
+
+    public Dto addGroup(String s) {
+      groups.add(s);
+      return this;
+    }
+
+    public List<String> getGroups() {
+      return groups;
+    }
+  }
+
+  private static final String SQL_TEMPLATE = "SELECT " +
+    "  project_authorization.project as project, " +
+    "  project_authorization.login as login, " +
+    "  project_authorization.permission_group as permission_group, " +
+    "  project_authorization.updated_at as updated_at " +
+    "FROM ( " +
+
+    // project is returned when no authorization
+    "      SELECT " +
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Define a constant instead of duplicating this literal "      SELECT " 4 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1192)